### PR TITLE
8332686: InetAddress.ofLiteral can throw StringIndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by AlekseiEfimov**
The proposed change fixes `InetAddress.ofLiteral` and `Inet6Address.ofLiteral` to throw `IllegalArgumentException` instead of `StringIndexOutOfBoundsException` when an empty IP addres literal string is specified.   
The `Inet4Address.ofLiteral` correctly throws `IllegalArgumentException` when a literal cannot be parsed as an IPv4 address literal. It was addressed before in [JDK-8315767](https://bugs.openjdk.org/browse/JDK-8315767) fix.

Testing: the modified `OfLiteralTest` test and other tests from `jdk-tier1` to `jdk-tier3` tiers

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332686](https://bugs.openjdk.org/browse/JDK-8332686): InetAddress.ofLiteral can throw StringIndexOutOfBoundsException (**Bug** - P4)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22518/head:pull/22518` \
`$ git checkout pull/22518`

Update a local copy of the PR: \
`$ git checkout pull/22518` \
`$ git pull https://git.openjdk.org/jdk.git pull/22518/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22518`

View PR using the GUI difftool: \
`$ git pr show -t 22518`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22518.diff">https://git.openjdk.org/jdk/pull/22518.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22518#issuecomment-2514958264)
</details>

### CODE REVIEW

File: test/jdk/java/net/InetAddress/OfLiteralTest.java [hunk 1]
```diff
@@ -373,7 +373,12 @@ private static Stream<Arguments> invalidLiteralArguments() {
                 Arguments.of(InetAddressClass.INET4_ADDRESS_POSIX, ""),            // empty
                 Arguments.of(InetAddressClass.INET4_ADDRESS_POSIX, "0x1FFFFFFFF"), // 2^33 - 1 is too large
                 Arguments.of(InetAddressClass.INET4_ADDRESS_POSIX, "0x100000000"), // 2^32 is too large
-                Arguments.of(InetAddressClass.INET4_ADDRESS_POSIX, "040000000000")
+                Arguments.of(InetAddressClass.INET4_ADDRESS_POSIX, "040000000000"),
+
+                // Empty literals
+                Arguments.of(InetAddressClass.INET_ADDRESS, ""),
+                Arguments.of(InetAddressClass.INET4_ADDRESS, ""),
+                Arguments.of(InetAddressClass.INET6_ADDRESS, "")
```
[the remark this answers is not in the document]
Comment (AlekseiEfimov) [reply test/jdk/java/net/InetAddress/OfLiteralTest.java:381]: Thank you for a good suggestion Jaikiran - I will add tests for blank literals now.

**Comment by dfuch** [comment]
Looks fine but you could additionally add this in InetAddress::ofLiteral:

```
      public static InetAddress ofLiteral(String ipAddressLiteral) {
          Objects.requireNonNull(ipAddressLiteral);
+         if (ipAddressLiteral.isEmpty()) {
+              throw IPAddressUtil.invalidIpAddressLiteral(ipAddressLiteral);
+         }

**Review by dfuch — APPROVED** [review]

**Review by jaikiran — APPROVED** [review]
Thank you for the update Aleksei. This looks good to me.

**Pull request closed without merging**